doc/user.xml

```diff
@@ -2388,6 +2388,19 @@ run</programlisting>
           <filename>mpc add smb://servername/sharename/filename.ogg</filename>
         </para>
       </section>
+
+      <section id="youtube_dl_input">
+        <title><varname>youtube-dl</varname></title>
+
+        <para>
+          Plays songs using <application>youtube-dl</application>. Use the <link
+          linkend="youtube_dl_playlist">youtube-dl playlist plugin</link>
+          instead of this directly, as it is able to populate more accurate
+          metadata. This plugin depends on curl and uses the <filename>
+          youtube-dl://</filename> URI scheme.
```

Comment:
Invented. Other applications that use `youtube-dl` such as mpv simply pass all http/etc URLs to it instead of using a specific URI scheme. If it fails to load, they move on to try other plugins like curl (or try curl first depending on user configuration). That approach is more user-friendly, and I'd like if it worked that way, but I'm not sure how well it fits into mpd's existing plugin design. `ytdl://` has my vote as it's slightly less annoying to type for a user.

Reply:
Can't we just make the plugin intercept all plain http:// YouTube URLs? That's what the user already has.

Reply:
We certainly can, and I think that will work best! My concern was that it may interfere with the curl plugin, since I don't see any explicit support for falling back to another plugin, but that could instead be done with `OpenCurlInputStream(uri)` if running the `youtube-dl` fails.

Reply:
It doesn't interfere because CURL wouldn't be able to get meaningful data from a `http://youtube.com/...` URL.

There is a "fallback" - the MPD core iterates over all plugins and the first one which doesn't return `nullptr` wins: https://github.com/MusicPlayerDaemon/MPD/blob/master/src/input/Open.cxx#L40

Yes, a whitelist would probably be good. Can it be auto-generated at runtime somehow from calling youtube-dl once?

Reply:
See [How can I detect whether a given URL is supported by youtube-dl?](https://github.com/rg3/youtube-dl/#how-can-i-detect-whether-a-given-url-is-supported-by-youtube-dl) It doesn't seem like we could generate it, but I think a whitelist also makes sense even if we just populate it with some high-profile websites. Personally I'd be inclined to just set it to `.*` as long as the fallback behaviour is working, sounds like a config option to me.

The issue with the fallback is that curl plugin will return a stream for any valid URL, before it makes the request and validates that there's playable media at the end of it. Otherwise I'm not sure we would need a whitelist - but I think a whitelist will help, and if we do eventually fail, we can fall back to the original provided URI anyway.

Reply:
No catch-all!
The fallback is only used when the plugin decides it does not support the URI, not when the plugin fails to open the URI.
And don't consider interpreting `youtube-dl`'s failure to open something as "not supported"; it is an error.
There must not be overhead for "regular" HTTP streams just because your plugin spawns `youtube-dl` each time.

Reply:
> There must not be overhead for "regular" HTTP streams just because your plugin spawns youtube-dl each time.

Agreed, which is why the whitelist must exist.

> And don't consider interpreting youtube-dl's failure to open something as "not supported"; it is an error.

Hm, I actually object to this, because as a user I know that every URL I use is supported by youtube-dl, and am okay with the rare said overhead for "regular" streams if I opt into it. Although I don't think it needs to be the default behaviour, it's reasonable behaviour that works like other tools I'm used to such as mpv, and not having the option to do that would be a bit off-putting.